dialogue-blocking-channels/src/main/java/com/palantir/dialogue/blocking/DefaultCallingThreadExecutor.java

```diff
@@ -0,0 +1,150 @@
+    }
+
+    @VisibleForTesting
+    DefaultCallingThreadExecutor(QueueTake queueTake) {
+        queue = new Queue(queueTake);
+        /** Notification when main future completes. */
+        Runnable notifier = queue::poison;
+        callback = new FutureCallback<Object>() {
+            @Override
+            @SuppressWarnings("FutureReturnValueIgnored")
+            public void onSuccess(Object _result) {
+                queue.submit(notifier);
+            }
+
+            @Override
+            @SuppressWarnings("FutureReturnValueIgnored")
+            public void onFailure(Throwable _throwable) {
+                queue.submit(notifier);
+            }
+        };
+    }
+
+    DefaultCallingThreadExecutor() {
+        this(BlockingQueue::take);
+    }
+
+    @Override
+    public ListenableFuture<?> submit(Runnable task) {
+        return queue.submit(task);
+    }
+
+    @Override
+    public void executeQueue(ListenableFuture<?> await) {
+        Preconditions.checkState(Thread.currentThread().getId() == threadId, "Executing queue on different thread");
+        Futures.addCallback(await, callback, DialogueFutures.safeDirectExecutor());
+
+        try {
+            RunnableFuture<?> toRun;
+            while ((toRun = queue.getWork()) != null) {
+                toRun.run();
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            abortQueue();
+
+            await.cancel(DO_NOT_INTERRUPT);
+        }
+    }
+
+    private void abortQueue() {
+        queue.poison();
+
+        RunnableFuture<?> toRun;
+        while ((toRun = queue.poll()) != null) {
+            toRun.cancel(DO_NOT_INTERRUPT);
+        }
+    }
+
+    private static final class Queue {
+        private boolean poisoned = false;
+        private final BlockingQueue<RunnableFuture<?>> queue = new LinkedBlockingQueue<>();
+        private final QueueTake queueTake;
+
+        Queue(QueueTake take) {
+            this.queueTake = take;
+        }
+
+        public synchronized ListenableFuture<?> submit(Runnable task) {
+            checkNotPoisoned();
+            ListenableFutureTask<?> future = ListenableFutureTask.create(task, null);
+            queue.add(future);
+            return future;
+        }
+
+        public synchronized void poison() {
+            poisoned = true;
+        }
+
+        public RunnableFuture<?> getWork() throws InterruptedException {
+            if (!isPoisoned()) {
+                return queueTake.take(queue);
+            } else {
+                return queue.poll();
+            }
+        }
+
+        public RunnableFuture<?> poll() {
+            return queue.poll();
+        }
+
+        private synchronized void checkNotPoisoned() {
+            if (poisoned) {
+                log.info("Submitted task after queue is closed");
+                throw new RejectedExecutionException("Queue closed");
```

Comment:
Is this always logged after an interruption? abortQueue calls queue.poison followed by future.cancel, which sets the future state to failed/canceled and executes the listener to submit the notifier to the poisoned queue?

Reply:
Good point.

Reply:
Done